### PR TITLE
8256516: Simplify clearing References

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this simplification of jlr.Reference clearing by VM code.

The function java_lang_ref_Reference::set_referent_raw was being used to
clear the referent of Reference objects, and only for that purpose.  This
change replaces that function with java_lang_ref_Reference::clear_referent,
which is much more obvious in intent.  That change is then percolated up
through callers in the obvious way.

Testing:
mach5 tier1

/label hotspot

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256516](https://bugs.openjdk.java.net/browse/JDK-8256516): Simplify clearing References

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to 84dc63e4166da7c11987255beb4e286a002a8267
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 84dc63e4166da7c11987255beb4e286a002a8267
 * [Per Liden](https://openjdk.java.net/census#pliden) (@pliden - **Reviewer**) ⚠️ Review applies to 84dc63e4166da7c11987255beb4e286a002a8267
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 84dc63e4166da7c11987255beb4e286a002a8267

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1286/head:pull/1286`
`$ git checkout pull/1286`

### CODE REVIEW

File: src/hotspot/share/gc/shared/referenceProcessor.hpp [hunk 1]
```diff
@@ -120,14 +119,8 @@ class DiscoveredListIterator {
...
-  // Make the referent alive.
-  inline void make_referent_alive() {
-    if (UseCompressedOops) {
-      _keep_alive->do_oop((narrowOop*)_referent_addr);
-    } else {
-      _keep_alive->do_oop((oop*)_referent_addr);
-    }
-  }
+  // Apply the keep_alive function to the referent address.
+  void make_referent_alive();
```
Comment (shipilev) [inline src/hotspot/share/gc/shared/referenceProcessor.hpp:123]: I wonder if moving this from the `.hpp` to `.cpp` has performance implications for callers. Maybe move to `.inline.hpp`?

**Review by rkennke — APPROVED** [review]
Looks good to me! Thanks!

File: src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp [hunk 2]
```diff
@@ -316,7 +316,7 @@ void ShenandoahReferenceProcessor::make_inactive(oop reference, ReferenceType ty
     reference_set_next(reference, reference);
   } else {
     // Clear referent
-    reference_set_referent(reference, NULL);
+    reference_clear_referent(reference);
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp:319]: Now I am looking at this code and wonder if we could just inline `reference_clear_referent` and `reference_set_next` both in Shenandoah and ZGC code. Probably something for a followup.

**Review by shipilev — APPROVED** [review]
Looks good, with minor nit.

File: src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp (hunk 2 above)
Comment (pliden) [reply src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp:319]: I'd prefer to keep the `reference_*` helper functions as is in ZGC.

**Review by pliden — APPROVED** [review]
Looks good!

**Review by mlchung — APPROVED** [review]

File: src/hotspot/share/gc/shared/referenceProcessor.hpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/shared/referenceProcessor.hpp:123]: The reason for moving it out of the .hpp was of course that the
change to call java_lang_ref_Reference::referent_addr_raw needs to #include
javaClasses.inline.hpp.

I don't see this move having any measureable performance difference, and not
even sure what the sign of any change might be.  A better refactoring might
be to package up the common remove/make_referent_alive/move_to_next sequence.

Comment (shipilev) [reply src/hotspot/share/gc/shared/referenceProcessor.hpp:123]: Yeah, it is probably only affects the path that is infrequently taken, i.e. marking through the resurrected referents during weak reference processing, that is probably only finalizers. It is fine to have it in `.cpp`.

**Review by shipilev — APPROVED** [review]

**Comment by kimbarrett** [comment]
/summary Provide and use explicit referent clearing instead of set to null.

**Comment by kimbarrett** [comment]
Thanks for reviews @rkennke, @shipilev, @pliden, @mlchung

**Pull request closed without merging**